random/seeded.ts

```diff
@@ -0,0 +1,220 @@
+
+/**
+ * The maximum value of a 64-bit unsigned integer plus one.
+ * `n % U64_CEIL` simulates uint64 integer overflow in C.
+ */
+const U64_CEIL = 2n ** 64n;
+/**
+ * The maximum value of a 32-bit unsigned integer plus one.
+ * `n % U32_CEIL` simulates uint32 integer overflow in C.
+ */
+const U32_CEIL = 2n ** 32n;
+/**
+ * The maximum value of a 64-bit unsigned integer.
+ * `n & U64_MAX` coerces any bigint `n` to fit inside a uint64.
+ */
+const U64_MAX = U64_CEIL - 1n;
+
+// arbitrary odd value (taken from example at https://en.wikipedia.org/wiki/Permuted_congruential_generator)
+const INITIAL_INC = 1442695040888963407n;
+
+/**
+ * A 2-tuple state for a {@linkcode SeededPrng}.
+ */
+export type State = [state: bigint, inc: bigint];
+
+/**
+ * Options for {@linkcode SeededPrng}.
+ */
+type SeededPrngOptions = {
+  /**
+   * The seed for the random number generator.
+   *
+   * - If a `bigint`, the seed will be used to initialize the state and increment.
+   * - If a `State` 2-tuple, the first element will be directly used as the state and the second as the increment.
+   */
+  seed: bigint | Readonly<State>;
+};
+
+/**
+ * A seeded pseudo-random number generator, implementing the
+ * {@link https://en.wikipedia.org/wiki/Permuted_congruential_generator | PCG32}
+ * algorithm.
+ *
+ * @example Usage
+ * ```ts
+ * import { SeededPrng } from "@std/random";
+ * import { assertEquals } from "@std/assert";
+ *
+ * const prng = new SeededPrng({ seed: 3377567226465013078n });
+ *
+ * assertEquals(prng.random(), 0.34011089405976236);
+ * assertEquals(prng.random(), 0.6603851807303727);
+ * assertEquals(prng.random(), 0.4863424440845847);
+ * ```
+ */
+export class SeededPrng {
+  #state: bigint;
+  #inc: bigint;
+
+  /**
+   * Constructs a new instance.
+   *
+   * @param seed - The seed for the random number generator, which can be
+   * either a 2-tuple of bigints or a single bigint, which
+   * will be converted to a state 2-tuple.
+   *
+   * @example Usage
+   * ```ts no-assert
+   * import { SeededPrng } from "@std/random";
+   * const prng = new SeededPrng({ seed: 14614327452668470620n });
+   * ```
+   */
+  constructor({ seed }: SeededPrngOptions) {
+    this.#inc = -1n;
+    this.#state = -1n;
+
+    this.state = typeof seed === "bigint"
+      ? SeededPrng.#seedToState(seed)
+      : seed;
+
+    // For convenience, allowing destructuring and direct usage in callbacks
+    // equivalently to Math.random()
+    this.random = this.random.bind(this);
+  }
+
+  /**
+   * The state for the random number generator.
+   * @returns The state as a 2-tuple.
+   *
+   * @example Usage
+   * ```ts
+   * import { SeededPrng } from "@std/random";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const prng = new SeededPrng({ seed: 14614327452668470620n });
+   * assertEquals(prng.state, [13062938915293834817n, 10846994826184652623n]);
+   * ```
+   */
+  get state(): Readonly<State> {
```

Comment:
Does `state` need to be accessible outside the class? I only see it for testing purposes.

Reply:
I think `state` should definitely be gettable, as it allows persisting a PRNg's current state for resumption later on. It's arguable whether or not it needs to be settable though, perhaps setting it should only be possible via the constructor or a `from` static method.

As with seeding, I'm still not too settled on how this should be exposed, maybe it should have explicit `state` and `inc` properties (and probably an `algorithm` property too, even though there's only one currently implemented).